### ISSUE TITLE
Brute-force protection: only download password-protected paste in case the receiver knows the password

### ISSUE THREAD

**Issue by rugk** [PrivateBin/PrivateBin#245]
Idea by @aldem in https://github.com/PrivateBin/PrivateBin/issues/174#issuecomment-326947154:
> When paste is created:
>
> 1. Generate HMAC from random salt, paste key & password
> 2. Store nonce & HMAC with paste
>
> When paste is requested:
>
> 1. Server sends salt to the client
> 2. Client computes HMAC using provided salt, known paste key & password, sends it to the server.
> 3. If stored HMAC matches, return paste and delete it.
> 
> Server still has no information which could be used to decrypt message, while it is able to verify that the client is really the intended recipient - thus it is safe to delete the paste.

**Note:** This is also a brute-force protection as an attacker cannot just download the paste and brute-force it.

**Comment by elrido** [comment]
One suggestion: Use a different salt for the client side HMAC then the server side one, otherwise any client can generate valid delete tokens. So we would have an internal "secret" server salt and a "public" one offered to clients for such operations.

**Comment by rugk** [comment]
Actually in this case we do not even have to call it "salt", but maybe just "token" or string. Maybe we also add the paste ID there.

So to expand what was written in the first post in a more formal way, we'd do:
``` 
Our syntax: HMAC(message, key)
HMAC(pasteID + pasteToken, pasteHashDecryptionKey + paastePassword)
```

The sever stores the finished HMAC and compares it.

As for the client side, this introduces one more JSON request. Here is the flow when requesting a paste:
1. HTML site downloaded. In the future this will be 100% static…
2. JSON request for pasteToken (or how we name it, paste salt).
3. Client calculates HMAC.
4. The "usual" JSON request for paste content, now only with HMAC as "authentication".

This will also prevent accidental paste deletions of "self-destroying pastes", without the correct hash and password.

**Comment by rugk** [comment]
Technically, we need a lib for that. I looked at the doc and I saw that HMACs are not provided by the Web Cryptro API. :cry: So https://github.com/Caligatio/jsSHA/ might be a way.

**Comment by elrido** [comment]
What's this then? https://github.com/diafygi/webcrypto-examples#hmac :wink:

I would definitely avoid introducing any new crypto/encoding (apart from pako, until there is mainstream browser support for deflate) libraries at this point. We had bad enough experiences with SJCL, rawdeflate and Base64.js in the past. It's tempting to use them, but it's better to stick to the built ins of the browsers, much less technical debt and better tested and automatically updated.

**Comment by elrido** [comment]
PS: After thinking about this in the terms of #347, I start to really like this HMAC idea. The paste would only need to returned by the API if it is either an old format without HMAC or if a matching HMAC got passed by the client.

If the wrong HMAC is passed, we could even consider rate limiting such a client, to avoid brute forcing paste retrievals by guessing the keys. Would do it with maybe 1s on the first failure and then increase the limit in a logarithmic scale on subsequent failures.

**Comment by rugk** [comment]
Problem for @elrido: In whatever way we do this, we may leak the fact that a paste is password-protected to the server.

As when a paste is not, you cannot verify any HMAC, or you verify an HMAC with an empty password (and that's easy to guess [/"brute-force"] for a server).

**Comment by rugk** [comment]
Also, we cannot download the paste to test decrypting without a password, because in terms of https://github.com/PrivateBin/PrivateBin/pull/216 this would delete a password.

**Comment by elrido** [comment]
HMAC: I assumed the HMAC would be ID (known to server) + 256 bit key + password (may be blank). The key can be brute forced as well with or without HMAC, can't it? And I assume we would always use the HMAC, otherwise the server would learn if the paste has a password.

Download to test: Full ACK. Hence the need for the HMAC.

**Comment by rugk** [comment]
> The key can be brute forced as well with or without HMAC, can't it?

* password
But, yes, of course.

> And I assume we would always use the HMAC, otherwise the server would learn if the paste has a password.

Thought that password is an empty string and thus can easily be guessed by the server and it can calculate the HMAC itself and verify with the data it knows, that this is the "password", i.e. password = no password used.

However, we may prevent that if we use the paste hash in the HMAC, too. So all seems to be ok.

**Comment by elrido** [comment]
Quick recap of the API flow as I understand it from the above:
1. Paste creation as usual, server responds with paste ID, (delete)token and a (new) public, paste specific salt (unlike the internal one that it uses to validate a sent delete token).
2. Client generates the HMAC based on paste ID, paste key (the URL hash) and the paste password (if there is one), sends this to the server in a new API call, authenticated by the token (so that only the creator can do this. \
If a client doesn't send that second call, the paste just has no HMAC and therefore is always returned (behavior for old pastes as well)
3. Visitor visits a URL and either:
   1. Visitor just fetches the paste, server returns it if it hasn't expired (just like now, this is to support third party clients), flow ends at this step
   2. Visitor has a client that supports the check API call, proceed to next step
4. Visitor calculates HMAC without password, sends that to check on the server using the check API call
5. Server responds with either:
   1. Failure, because a password is required or the HMAC is otherwise incorrect (brute force), proceed to next step
   2. Success, and sends the paste alongside, flow ends here
6. Visitor gets asked for a password, new HMAC is calculated based on this input and sent to the server, loop back to step 5.

Please review and send your corrections.

**Comment by elrido** [comment]
Addendum: Had a think about if it's possible to avoid having to send two requests during the creation.

The problem there is that the creator doesn't know the paste ID, as that gets generated server side. Here is an altered proposal that would address that:
1. Paste creation as usual, but client also sends along an HMAC based on the paste key (the URL hash) and the paste password (if there is one), server responds with paste ID and (delete)token, server internally generates a new HMAC based on the paste ID and the creators HMAC and stores this along the paste.
If a client doesn't send an HMAC, the paste just has no HMAC and therefore is always returned (behavior for old pastes as well).
3. Visitor visits a URL and either:
   1. Visitor just fetches the paste, server returns it, if it hasn't expired (just like now, this is to support third party clients), flow ends at this step.
   2. Visitor has a client that supports the HMAC, proceed to next step
4. Visitor calculates HMAC of paste ID and HMAC of key and empty password, sends that to check on the server using the usual fetch API call
5. Server responds with either:
   1. If the paste is a burn-after-reading paste, it gets sent in any cas and deleted on the server right awaye (this addresses the concern in #347). If the decryption fails, the visitor gets asked for a password, but the paste is lost if the browser tab is closed, same as now.
   2. Failure, because a password is required or the HMAC is otherwise incorrect (brute force), proceed to next step
   3. Success, and sends the paste alongside, flow ends here
6. Visitor gets asked for a password, new HMAC is calculated based on this input and sent to the server, loop back to step 5.

**Comment by rugk** [comment]
ehm… Or we could simplify that if we just don't use the paste ID in the HMAC? ~~I mean, the paste encryption key should be unique enough.~~

Actually, we only talk about about `pasteToken`, so:
```
HMAC(message, key)
HMAC(pasteID + pasteToken, pasteHashDecryptionKey + paastePassword)
```
would get
```
HMAC(pasteToken, pasteHashDecryptionKey + paastePassword)
```

`pasteToken` is just randomly generated on the client and public data.
And randomly generated is always enough randomness, we hardly need to paste ID there as far as I see…

**Comment by elrido** [comment]
Ok, but then how is the pasteToken shared with the visitor? Would it be appended to the URL?

We also can't use the message, the visitor needs to be able to generate the HMAC from bits of information it already has (in the URL). And we can't have the visitor ask the server for that pasteToken, as this would definitely leak the information that the paste exists. Or we just return a fake token if the paste doesn't exists?

Aside: I think as there is the brute force aspect to this, we should probably use a key derivation instead of just an HMAC. We could use PBKDF2 again with the high iteration count. This should further ensure that (i.e. the server operator) can't use that information to brute force the key + password and it makes it also more costly for a visitor without password to trial and error them.

**Comment by rugk** [comment]
> And we can't have the visitor ask the server for that pasteToken, as this would definitely leak the information that the paste exists

This was the idea…

> Or we just return a fake token if the paste doesn't exists?

`0000000000` or so :wink:.

> I think as there is the brute force aspect to this, we should probably use a key derivation instead of just an HMAC. We could use PBKDF2 again with the high iteration count. 

Not really needed, as the server should do rate-limiting. That's (brute force protection) is what this whole issue is about…
So calculation speed does not really matter. It maters that we can prove to the server that the user knows a password without revealing the password to the server.
Also we include the paste key, so there is enough entropy, so it won't be possible to brute-force the HMAC for the server.

**Comment by elrido** [comment]
Sorry, now I am confused. Could you please share the amended API flow how you imagine it?

Reg. rate limiting - currently we only limit the creation, but not the fetching of pastes. I understand that you suggest to rate limit the fetching of pastes as well to prevent brute forcing of requests, correct?

**Comment by rugk** [comment]
Yep, this would be required.

The protocol is [already described here kinda](https://github.com/PrivateBin/PrivateBin/issues/245#issue-255067586). I later described it again with pasteId, but as we see this would make stuff quite complicated for creating the paste (as you've described), so better not use the paste ID.

**Comment by elrido** [comment]
Ok, so you would prefer to have two calls for retrieval. I'll do a PR with a simple implementation and we can then discuss further steps based on it.

**Comment by aldem** [comment]
It seems that problem is getting more annoying - more and more services, firewalls and antiviruses are peeking into URLs so burn-after-reading pastes are lost.

After some thinking and reading all discussions I still believe that my [original idea](https://github.com/PrivateBin/PrivateBin/issues/245#issue-255067586) is good enough, though there are few things that could be improved:
- HMAC must be replaced by PBKDF2 with (in)sane number of rounds
- Retrieval must be rate-limited
- Optionally: do not store salt on the server (more on this below)

Another option is to use SRP, but this adds unnecessary complexity without significantly improving security, while simple solution is "good enough" for most cases where it is needed - configurable, properly documented and disabled by default, it will not make PrivateBin less secure, especially in use cases when server is trusted.

With PBKDF2 and rate limiting in place (assuming server is well-protected and inaccessible to the attacker) there is virtually no chance that anyone that has access to the "retrieval token" will be able to brute force the password faster than the derived key so pastes are still safe.

If the show-stopper is untrusted server (or server admin) we could omit storing the salt and compute token solely from the known data (key and password), this way we don't even need to disclose to the server that paste is password-protected - token is always generated, for non-password case empty password is assumed.

When token (if provided) does not match, server responds "not found" and UI informs the user that "Paste does not exist, expired or password-protected" and offers "If you are certain that the paste is password protected, please enter the password and try again".

Though if the server is really untrusted, I have to quote @rugk:
> Especially given that a malicious server admin could have 1000 other ways for an attack (change JS), so we already assume some trust...

And now we have two options:
- Leave everything as is, limiting use cases
- Adopt simple and secure enough solution which allows more use cases without compromising security at least for trusted servers (like ones used by teams, user groups or companies internally).

Any comments?

**Comment by rugk** [comment]
> HMAC must be replaced by PBKDF2 with (in)sane number of rounds

Good point, but in reality we need to do all that anyway, so we could re-use the [derived key](https://github.com/PrivateBin/PrivateBin/blob/fb66e2392bc833b0e43ac8b553e587c06351a2c6/js/privatebin.js#L1056) we have there anyway, could not we?
The thing is just this has to stay private to the client, so then, a simple HMAC on that should be sufficient, so we don't send the decryption key to the server. :thinking: 

Okay, so if we wanted the server from brute-forcing the hash+password then yeah, we should propbably just do another key stretching rounds on that…

**Comment by aldem** [comment]
If we run HMAC on top of derived key, then probably this is sufficient, we could even cut it in half (probability of collision is very low anyway). PBKDF2 is indeed to ensure that server operator (if server is not trusted) could not brute-force it.

Though another idea crossed my mind... What if we will generate paste id on the client? We could run another PBKDF2 on top of key + password (different number of rounds and/or different algorithm and/or key length), and use it as the paste id - thus client knows in advance paste id, but anyone who wants to retrieve this paste must know the password - without the password server simply could not find it. Odds of collision are quite low, but even if this happens (when creating the paste) client simply generates new key and tries again.

**Comment by elrido** [comment]
If I may make a suggestion: Should we maybe model this in something like [Verifpal](https://verifpal.com/), first? The last time we worked on this we changed the implementation several times in the PR as we kept finding potential issues (PR #447 technically works, but we felt it was too risky). If we could formalize the protocol first, the implementation will be much more straight forward.

**Comment by rugk** [comment]
Well I looked at the video and that looks pretty damn cool! Formal verification in an somewhat easy way, ugh… nice.

Maybe make a new issue?

**Comment by aldem** [comment]
Verifpal is nice tool but there is nothing new to model here. Just look at this (simplified) protocol:

1. vt = HMACorPBKDF2(password, derived-key)
2. Now we have two options:
  a. Use vt (fully or partially) as paste id (instead of relying on the server to generate one)
  b. Store vt (fully or partially) as "verification token" and make the server check it on GET

Server does not know the the key nor the password - so it could not brute force the password thus no change in threat model for the server.

MitM does not have more information than it has now - it knows the key anyway, but it has to do a lot more to *retrieve* the paste if it is password protected, i.e. it *has* to brute-force the password *and* to query the server for *every* attempt, which slows it down drastically (especially if rate-limited).

The only (UI-wise) change is that when paste is not found (if we use 2.a) or token does not match (2.b) client must offer to enter the password and try again (alternative is to report condition "token does not match" - but this will disclose existence of the not-yet-retrieved paste, which may be not desirable).

What we *do* lose is the notification that (password-protected burn-after-reading) paste was retrieved by someone, though this could be worked-around - on successful retrieval (which is only possible if client knows both the key & password) we could mark the paste as "deleted" (by deleting its content) and report this fact by omitting password prompt.

Sure this could be bad as one could prove that paste *did* exist, though without content it does not mean much.

**Comment by elrido** [comment]
Please have a look at #447 and the discussion we had there (if not already done). It implements something like you describe, but we found that it is not as simple as that. I am happy work on it again, but only after we do get some modeling done first. If I do find some time over the long weekend, I'll give the verification thing a try and share the results with you here.

If we do get a positive result for the existing implementation I am happy to merge it after all and if flaws are found, then we have a model to review and work on those issues, before we waste time implementing something that ends up putting users at risk.

**Comment by elrido** [comment]
As promised, but later then anticipated, I did create those Verifpal models and they can be found and played with at https://github.com/PrivateBin/Verifpal

All of the questions fail verification - the paste is not confidential from even a passive attacker. This is of course not what the project aims to achieve, since the URL with the key and the password is passed around from creator to reader in a probably unprotected way or even published on a public site. The server is supposed to be protected from getting to know the contents of the paste, but I haven't figured out how this can be modelled in Verifpal.

I did model both the [implementation in the PR](https://github.com/PrivateBin/Verifpal/blob/master/challenge-response.vp) and the [nonce based one proposed by @aldem](https://github.com/PrivateBin/Verifpal/blob/master/challenge-response-nonce.vp) and have to say that I would feel more comfortable with the nonce guarding the key and password from the server, even if it makes the hash part of the URL longer. A model of the [current, basic protocol](https://github.com/PrivateBin/Verifpal/blob/master/paste.vp) is also included for comparison.

The syntax is straight forward and the resulting models pretty much human readable. Really helps (me, at least) to understand the flow of information.

**Comment by rugk** [comment]
First, awesome, thanks! I may look into that when I setup my own Verifpal here. (stupid snap package they have)

> All of the questions fail verification - […] but I haven't figured out how this can be modelled in Verifpal.

From what I know/seen in their intro video, you should be able to solve this by specifying a key as "guarded" (with `[]` brackets around it), that means, it is verified out-of-band somehow, i.e. as if you'd scan a QR code in signal. IMHO this is our paste hash, that we have to assume is shared secretly over some out-of-band channel. (This is an assumption [we always made](https://github.com/PrivateBin/PrivateBin/wiki/FAQ#if-i-use-privatebin-and-post-the-link-in-a-forum-in-social-media-etc-is-my-content-secure): if you share the link, it's public, anyone can read it; if you share it secrectly)

Also, I guess you should not include the server in your model. The server is our attacker, you already specify that with the first line.
The way where they model Signal etc., they also did not include the server, neither should we… :thinking: 
This is the e2e-level moddeling of course, that may only work with our current approach.

Because of course for challenge-response, Iwe have to take the server into account.

**Comment by elrido** [comment]
guarded values in Verifpal: I had tried it, but it makes no difference for our case, because (from the manual) "while an active attacker can still read it, they cannot tamper with it." - the passive attacker still reads the guarded key in transmission and can decrypt the paste. I like the modeling aspect of the language, but our protocol clearly is not protecting the secrets from the passive eavesdropper, only from the server and it seems that verifpal can't model this as such.

Storing the iv in clear text in the authenticated data alongside the encrypted data [should be safe](https://crypto.stackexchange.com/questions/2280/why-is-the-iv-passed-in-the-clear-when-it-can-be-easily-encrypted/2287#2287) and has been replicated from the SJCL library format.

Storing the _key derivation_ salt (server salt serves a different purpose and isn't shared outside the server) on the other hand doesn't grant the server that much more leverage to crack the password, they still had to brute force it, going through 100000 iterations per key + password combination. And the server doesn't want to do that in our threat model. If the server wants to crack your password, they can pass a nicely SRI-hashed JS file along the website that just extracts the key from the URL-hash and passes it on. We assume that the server wants to protect itself from that knowledge, as not to become legally responsible for hosted content, until notified of any problematic item. As the user you still have to be able to trust the server (operator) not to steal your key or password, regardless if the salt is stored on it or not.

**Comment by aldem** [comment]
I understand that storing iv/salt is not a threat, but if we can avoid this - why not? Deriving both from the key+password is as safe as producing random values, but we save a bit of space. Actually, iv is not needed at all since key is always random value.

As to Verifpal - yes, it is difficult to model PB protocol as there is no way to indicate secure channel, though this could be worked around by specifying "knows private key" and "knows password optional_secret" for creator/reader and removing "key" from all communications. However, this will also require to apply PW_HASH() to "optional_secret" or it will be deductible for an attacker (probably a glitch in Verifpal, as actual usage with KDF prevents this).

I will try to play around with Verifpal models but most likely I would not have enough time before weekend...

**Comment by elrido** [comment]
We should avoid re-using the key for some other purpose then as the secret in the KDF as that would weaken the KDF or the crypto - these really need to be separate, cryptographically strong random numbers. We would have to add additional pieces of information to the URL if we don't want to store them on the server, making it longer. Also keep in mind that changing such fundamental properties breaks the existing paste support and would have to be carefully implemented as a new version of the paste format so that the older pastes / URLs still work. This is pretty tricky to get right as you can see in the release history around version 1.3.

Lets stick to the topic at hand please - looking forward to see your improvements for  the verifpal models - I sent you an invite for that repo.

**Comment by jsarenik** [comment]
Hello everyone!

Excuse me please for not reading all the posts here, but let me explain a way I thought about (and was happy to find an already-living issue with the same topic) so here it is:

 1. User creates a "Burn after reading" paste.
 1. A random token is generated.
   i. Hash of this random token is stored on server in a publicly non-accesible path.
   ii. The token itself is added into paste's JSON and encoded the same way as comments.
 1. When another user opens the paste and uses a valid password the token mentioned above is read, decrypted and then a GET is sent to a particular endpoint which will make server delete the file. In case of normal (non-burn) paste, the endpoint can increment the number of successful reads of the paste.

Then another thing comes to my mind: There should be fake content and on clientside JS processing time connected with non-existing pastes. That would make it impossible to distinguish whether the paste exists or not.

What would you say about a separate issue for Verifpal? To me it seems that this particular problem can be cherry-pick-fixed even without big things like Verifpal.

**Comment by jsarenik** [comment]
After re-reading the issue I realized my comment was off-topic here. Sorry.

**Comment by rugk** [comment]
Remember to see https://github.com/PrivateBin/PrivateBin/issues/620 for an alternative implementation.

**Comment by elrido** [comment]
This is an attempt to model the protocol proposed in #1055 - please comment on any flaws or misunderstandings in this:

1. if the paste is created as burn-after-reading AND with a password:
   1. store the burn-after-reading paste as we do now
   2. store a second paste with a to-be-defined, new formatter (i.e. "pasteuri"), with burn-after-reading and comment flag set to false, but otherwise the same expiration and password, but newly generated IV, salt and key (these latter parameters should always be different between two pastes and should never be re-used). The message stored in this paste is the URL including key that would otherwise be displayed to the creator as the paste URL.
   3. we present the creator with a paste URL including key to the second paste
2. when a visitor's client has decrypted a paste and the new format is detected:
   1.  the client loads the paste ID from this new URL (which triggers that paste to be deleted, when it is flagged as burn-after-reading)
   2. the client decrypts this new paste with the same password and the new key (this could still fail, if that paste link wasn't created as per the above process) and gets this paste displayed instead

Challenges with this scheme I can think of:
- By default we employ a rate limiter that prevents clients from creating more then one paste or comment per 10 seconds. If we don't (want to) change or weaken this, the client would simply have to retry the creation of the second paste several times until it works (the client doesn't know the configuration of the rate limiter on the server, it could be above or less than 10 seconds). Or we could ignore a rate limit, but only if the stored item is a paste with this new formatter set - this would introduce a denial of service risk. Or we do change the API to allow the creation of the paste pair in one request (see below).
- I can't immediately construct a scenario, but this probably introduces an information leak with this new type of paste. While the burn-after-reading paste itself would still get destroyed, the second paste would get left behind as the server doesn't have any connection between it and the burn-after-reading one. Maybe the burn-after-reading paste itself would need to be extended as well, so it can include a secondary paste ID to be deleted when it gets deleted? The paste ID is generated server side, so maybe we would need to extend the API after all, allowing a pair of pastes to be created at the same time, allowing the server to store the ID of the second paste with the first one (for deletion) and the ID of the first paste in the second paste (unencrypted - the server doesn't know the key and password) for redirection purposes. In this scheme the only encrypted information in the second paste is the key for decrypting the first paste.
- Should we consider storing the full paste URL or only paste ID and key in the second paste? The full URL would allow for having these two items stored on different instances. But if we do want to tie these two pastes together on the server side to allow the server to delete the second paste once the first gets burned, then they have to live on the same instance and storing the URL is moot.
- While this doesn't require a change in the API, it changes the format of the paste, so at least for the second paste type (with the new formatter setting) we should use a new paste format version (i.e. `v: 3` ?) so that old clients (including 3rd party ones) can display an appropriate error message if they have no support for displaying this.

Any other flaws anyone can think of?

**Comment by rugk** [comment]
Don't get why we would need to store two pastes. The idea seems to come from https://github.com/PrivateBin/PrivateBin/issues/1055, but AFAIK a real protocol could achieve the same thing.

**Comment by elrido** [comment]
> Don't get why we would need to store two pastes.

It's about the burn-after-reading pastes with password - The second paste is set to burn-after-reading, but the first is not. So if someone fails to decrypt the first paste, they don't get the URI to the second one, so it won't get burned.

This is less elegant than the approach with a different protocol that we implemented but abandoned due to the risks it introduced. This solution still guarantees that the actual burn-after-reading paste only can be opened once - the first paste serves as a trial-and-error mechanism for the decryption with password.

**Comment by rugk** [comment]
I'd consider that really less elegant due to the fact of the server rate limiter… this is a major obstacle in that context, one cannot wait just for the limit to end…

**Comment by rugk** [comment]
> Another option is to use SRP,

This was another option [proposed here](https://github.com/PrivateBin/PrivateBin/issues/245#issuecomment-608709477) that got lost somehow. Yeah more complex, but the [_Secure Remote Password protocol_](https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol) seems to have exactly the properties we want (plus some more?): for a server is is (nearly) zero-knowledge, i.e. the server does not know the password, but can verify it. And I found a Mozilla JS implementation in https://github.com/mozilla/node-srp or better the modern https://github.com/LinusU/secure-remote-password.
If we build upon existing standards, we have to model not so much hmm…

Looking at the last link it indeed seems to be quite a complex handshake hm…
And we would need the server-side, too: [Aka this PHP one](https://github.com/artisansdk/srp).

Dunno…

**Comment by elrido** [comment]
I'd feel much more comfortable implementing a standard protocol.

From reading into it, the registration step could be integrated into the paste creation request. new paste format, so also new version - now the server knows there is some extra step to do when retrieving this paste.

What would have to change is paste retrieval:
1. the client tries to fetch the paste the current way, as it doesn't yet know if it needs a password or not.
2. if the paste is the current format, the server can hand it out as usual (burning the paste if necessary), but if it's the SRP type, it must indicate with a new status (password required, I think there is a HTTP status for that we could use).
3. the client then can ask the user for the password, initiate an SRP challenge on a new API, the identity is the paste ID, in our case
4. the server assigns the client a PHP session responds with it's side of the challenge, stores temporary results in session
5. the client completes the authentication (and sends it's session cookie along)
6. the server verifies the clients authentication, if it succeeds, sends along the paste (burning the paste if necessary) and can in either case cleanup the session on it's end
7. the client should cleanup it's session
8. if step 6 failed, it can ask the user for a new password and return to step 3

Optional: For burn-after-reading pastes it may now be of interest to track number of failed password challenges on the server side and pass it along with the paste upon success, so the user can be told "x failed password attempts before opening this paste" or such.

Does this sound about right? Anything I missed?

**Comment by rugk** [comment]
@elrido one thing to keep in mind: if we add an additional json request or rather much more as the protocol requires we of course leak meta data, I.e. that it is a SDP and obviously the number of failures/attempts.

Apart from that this sounds good.

**Comment by rugk** [comment]
Reopening as origin issue has not been solved yet.

**Comment by aldem** [comment]
I am back to the party (took just few years)... :)

Unfortunately, VerifPal _does not_ explicitly allows to declare "secure channel" (invisible to passive attacker), so obviously the model by @elrido fails verification because it _explicitly_ specifies that they `key` and `optional_secret` are _observable_:

`Creator -> Reader: paste_id, key, optional_secret`

Any clause like this makes *all constants* visible to the attacker, no matter if you define phases or not, as by definition the attacker has access to all _messages_. Phases are only meaningful if we use `leaks` and don't expose constants in messages.

Thus, to make things _realistic_, we have to declare everything that is _not sent over the network_, or otherwise invisible to the attacker, as `knows private ...`.

Once we do this, and only describe visible traffic, everything perfectly verifies (as expected), including my proposal (which is still, I believe, is the simplest and secure way).

My model: https://gist.github.com/aldem/b34832d7e0227dc719a05afb6d0fce7f

Play with it, and try to find a flaw.

**Comment by rugk** [comment]
BTW did some quick research again whether there are modern zero knowledge/challenge response libraries that could help here and found nothing fancy. Only:
* Etebase (from the [EteSync](http://www.etesync.com/) project) seems to use some public key and stuff for authentication (but technically does what we want) https://docs.etebase.com/protocol-specs/authentication
* found a lib https://suwasto.github.io/kmpzkp/ for zero-knowledge proofs based on [Schorr's algorithm](https://en.wikipedia.org/wiki/Schnorr_signature)

**Comment by dillonstreator** [comment]
Hi all — I’m the creator of crypt.fyi (inspired in part by this project). It’s a zero-knowledge secret sharing platform that addresses this problem. Thought I'd share my approach since it's been running in production and might be useful for the discussion.

I avoided SRP / complex challenge–response and use a simple hash-based proof of knowledge instead.

Creation
- Client generates a 256-bit random encryption key
- Computes `SHA-512(key + password)` (password can be empty)
- Sends encrypted content + hash to server
- Key stays in the URL fragment (never sent)

Retrieval
- Client submits hash after locally computing
- Server compares and only returns ciphertext on match
- Decryption is fully client-side

Security properties
- The 256-bit random key provides sufficient entropy that brute-forcing the hash is infeasible, even if the password is empty.
- The hash is always computed the same way, so the server cannot distinguish between “password-protected” and “not password-protected” vaults.
- Verification is strictly proof-of-knowledge of `key + password`; the server never sees either value.

I also implemented a few safeguards that (for my threat model) reduce the need for a more complex PAKE-style protocol:
- **Failed-attempt limit** (`fc`): auto-burn after N bad attempts (1–10 configurable)
- **Read-count limit** (`rc`): allow N successful reads before burn
- **Rate limiting**: 10 req/min/IP by default

Because attempts are both rate-limited and bounded, online guessing is tightly constrained, which made a simpler construction acceptable for this use case.

Optional webhooks are also available for reads, failures, and burns.

Spec + implementation details:
https://github.com/osbytes/crypt.fyi/blob/main/SPECIFICATION.md#23-zero-knowledge-proof-mechanism

One trade-off I accepted:
`400` for wrong hash vs `404` for non-existent vault (so existence can be inferred). I considered that acceptable given 20-character random vault IDs, better UX for users, and the failed-attempt limit burning secrets after N tries.

Happy to discuss further — and thanks for all the work on this project.

**Comment by aldem** [comment]
@dillonstreator Actually, this (HMAC over key + password) was the original idea which opened this issue about 8 years ago :)

Anything beyond HMAC (or reasonable complexity) is unnecessary overcomplication which brings no "added security", because the only way to recover a 256-bit random key from a 256+-bit secure hash is the brute force, which is unfeasible even in quantum era.

**Comment by elrido** [comment]
I would not want to risk using just a straight sha256 or sha512 hash, as at least partial rainbow tables for these do already exist ([example](https://sha512.web-max.ca/)). Sure, 32 random bytes are likely still not in there, but using PBKDF2 with 100'000 rounds or such does protect against that and is the same algorithm we already use for key derivation for the encryption.

Why I abandoned that merge request was the realization that even with key and password reversed or other gyrations, we would have ended up storing something that is very close to the key + password on the server, which might let someone with server access get at it. My latest thinking on this was to use the [merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) technique, were the different known bits (maybe two halfs of the key or key plus password if one exists) are hashed, then a hash or PBKDF2 across is formed and given to the server. Anything that makes it nearly impossible for the server side to get at the key from that stored secret.

Also, when passing this token to the server to retrieve the document, it also is sent over the wire where third party eves-droppers can also acquire it and then use it to circumvent the brute-force protection. Protecting against that would again require some protocol and creation of a shared secret (i.e. [DH](https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange)) to protect the token exchange with.

**Comment by aldem** [comment]
TL;DR As long as we keep our _key_ secret (this implies no side-channel or other leaks), there is no known way for an attacker to decipher the paste without brute force, with or without hash (assuming secure hash), and any kind of brute force over 256 bit key space is unfeasible even with generally available quantum computing. We can even make all pastes and hashes publicly available - this won't change anything (besides we can't deny their existence), they are unbreakable.

Something as simple as `SHA256(key || password)` is more than enough (= no shortcuts, as SHA-256 is preimage resistant) and quantum-resistant - because we have 256-bit _random_ key, which can only be found by brute force, which requires at least 2^128 operations (only if we have fault-tolerant quantum computer and use Grover's algorithm).

Even any kind of KDF is overkill here - this is only relevant if we derive the key from relatively low-entropy passwords, but our key material is _random_ data already.

> which might let someone _with server access_ get at it

If someone got so far, then, most likely, the game is over already - the served JS can be modified in a way that reveals everything that someone wants.

> where third party eves-droppers can also acquire it and then use it to circumvent the brute-force protection

What exactly is "circumvented"? If a client (user) sends a valid hash or SRP token (or whatever else is used), then the paste (in encrypted form) is _already_ observable - there is nothing else to "circumvent". However, hash gives an attacker absolutely no advantage - it would, if we would rely on simple hashed (and unsalted) passwords.

If an attacker knows only paste-id - then they *still* need to find the proper hash first to retrieve the content - which sends them back to brute-force over 256-bit key space.

The thing is - DH, RSA or whatever else more complicated (including all forms of SRP) adds _nothing_ to security - because we already have quantum-resistant symmetric cipher with 256 bit _random_ key.

**Comment by rugk** [comment]
Hmm tend to agree here. Even the famous HaveIBeenPwned apparently uses [SHA-1 hashes of passwords to be checked](https://www.troyhunt.com/understanding-have-i-been-pwneds-use-of-sha-1-and-k-anonymity/) against their online DB and many services and browsers even implement their feature now. More explanation can be found [from reliable sources](https://blog.cloudflare.com/validating-leaked-passwords-with-k-anonymity/) like here. [Etc](https://duckduckgo.com/?q=k+anonymity+haveibeenpwned&t=fpas&ia=web). 

Now what they do is a little different as their use case is different:
* k anonymity is not really needed for our use case
* we have no plain passwords, if we combine them with our secret key in any case, yes that should be enough entropy for everything
* salting? Or random key, given it is random enough, should prove to be essentially that salt.

A compromise would be to not submit the full hash to the server but just a part of it, which would theoretically allow collisions aka to retrieve the paste data with a wrong key/password (note:just like it does currently!) but reduce the entropy submitted to the server? 

I guess what would help to build confidence here is to make a PR in https://github.com/PrivateBin/Verifpal and model the new implementation in Verifpal. Then we can be formally sure no vulnerabilities/other flaws are introduced.
